pkg/gardenlet/apis/config/types.go

```diff
@@ -0,0 +1,227 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"time"
+
+	gardencorev1alpha1 "github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	componentbaseconfig "k8s.io/component-base/config"
+	"k8s.io/klog"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// GardenletConfiguration defines the configuration for the Gardenlet.
+type GardenletConfiguration struct {
+	metav1.TypeMeta
+	// GardenClientConnection specifies the kubeconfig file and the client connection settings
+	// for the proxy server to use when communicating with the garden apiserver.
+	GardenClientConnection GardenClientConnection
+	// SeedClientConnection specifies the client connection settings for the proxy server
+	// to use when communicating with the seed apiserver.
+	SeedClientConnection componentbaseconfig.ClientConnectionConfiguration
+	// ShootClientConnection specifies the client connection settings for the proxy server
+	// to use when communicating with the shoot apiserver.
+	ShootClientConnection componentbaseconfig.ClientConnectionConfiguration
+	// Controllers defines the configuration of the controllers.
+	Controllers GardenletControllerConfiguration
+	// LeaderElection defines the configuration of leader election client.
+	LeaderElection LeaderElectionConfiguration
+	// Discovery defines the configuration of the discovery client.
+	Discovery DiscoveryConfiguration
```

Comment:
Can make it optional.
```suggestion
	Discovery *DiscoveryConfiguration
```